crypto/encode_decode/decoder_meth.c

```diff
@@ -466,42 +467,35 @@ int OSSL_DECODER_is_a(const OSSL_DECODER *decoder, const char *name)
     return 0;
 }
 
-struct decoder_do_all_data_st {
-    void (*user_fn)(void *method, void *arg);
+struct do_one_data_st {
+    void (*user_fn)(OSSL_DECODER *decoder, void *arg);
     void *user_arg;
 };
 
-static void decoder_do_one(OSSL_PROVIDER *provider,
-                           const OSSL_ALGORITHM *algodef,
-                           int no_store, void *vdata)
+static void do_one(int id, void *method, void *arg)
 {
-    struct decoder_do_all_data_st *data = vdata;
-    OSSL_LIB_CTX *libctx = ossl_provider_libctx(provider);
-    OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
-    const char *names = algodef->algorithm_names;
-    int id = ossl_namemap_add_names(namemap, 0, names, NAME_SEPARATOR);
-    void *method = NULL;
+    struct do_one_data_st *data = arg;
 
-    if (id != 0)
-        method = ossl_decoder_from_algorithm(id, algodef, provider);
+    data->user_fn(method, data->user_arg);
+}
 
-    if (method != NULL) {
-        data->user_fn(method, data->user_arg);
-        OSSL_DECODER_free(method);
-    }
+void ossl_decoder_do_all_prefetched(OSSL_LIB_CTX *libctx,
+                                    void (*user_fn)(OSSL_DECODER *decoder, void *arg),
+                                    void *user_arg)
+{
+    struct do_one_data_st data;
+
+    data.user_fn = user_fn;
+    data.user_arg = user_arg;
+    ossl_method_store_do_all(get_decoder_store(libctx), &do_one, &data);
 }
 
 void OSSL_DECODER_do_all_provided(OSSL_LIB_CTX *libctx,
                                   void (*fn)(OSSL_DECODER *decoder, void *arg),
                                   void *arg)
 {
-    struct decoder_do_all_data_st data;
-
-    data.user_fn = (void (*)(void *, void *))fn;
-    data.user_arg = arg;
-    ossl_algorithm_do_all(libctx, OSSL_OP_DECODER, NULL,
-                          NULL, decoder_do_one, NULL,
-                          &data);
+    (void)OSSL_DECODER_fetch(libctx, NULL, NULL);
```

Comment:
We should look into avoiding this call unless something has changed.  Other things are cached and flushed, this would be one more.

Doesn't have to be this PR.

Reply:
The core fetching routine has a pre-condition that does exactly that

Reply:
I think you missed the point: `OSSL_DECODER_do_all_provided()` is a _public_ function.  Calling it always incurs the fetching cost even though we know when it cannot have changed since the previous call (because we flush other caches in that event).

Reply:
I hear what you say, and then I'm staring at the following, wondering how that doesn't fulfill exactly what you're talking about?

https://github.com/openssl/openssl/blob/b19fcc66d382357617744690dc3363947de2cb6f/crypto/core_algorithm.c#L45-L57
https://github.com/openssl/openssl/blob/b19fcc66d382357617744690dc3363947de2cb6f/crypto/core_fetch.c#L27-L47
https://github.com/openssl/openssl/blob/b19fcc66d382357617744690dc3363947de2cb6f/crypto/provider_core.c#L1287-L1305